### PR TITLE
Add analysis to compute mappings between instructions and basic blocks.

### PR DESCRIPTION
**Pull request by dnovillo**
Note that this is currently running into some testsuite failures.  I believe these should be fixed by #957.

### CODE REVIEW

File: source/opt/ir_context.h [hunk 1]
```diff
@@ -25,10 +25,20 @@ namespace ir {
...
+  // 1. Make sure it gets invalidated by IRContext methods that add or remove IR
+  //    elements (e.g., KillDef, KillInst, ReplaceAllUsesWith).
+  //
+  // 2. Add handling code in BuildInvalidAnalyses and
+  //    InvalidateAnalysesExceptFor.
   enum Analysis {
     kAnalysisNone = 0x0,
     kAnalysisBegin = 0x1,
     kAnalysisDefUse = kAnalysisBegin,
+    kAnalysisInstrToBlockMapping,
```
Comment (s-perron) [inline source/opt/ir_context.h]: Since we are going up by powers of 2,should explicitly set to value to 0x2?

File: source/opt/ir_context.h [hunk 2]
```diff
@@ -25,10 +25,20 @@ namespace ir {
...
     kAnalysisEnd = 0x2
```
Comment (s-perron) [inline source/opt/ir_context.h]: This is suppose to be 1 past the end, so it should become 0x4.

File: test/opt/def_use_test.cpp [hunk 3]
```diff
@@ -1318,7 +1318,9 @@ TEST_P(KillInstTest, Case) {
       BuildModule(SPV_ENV_UNIVERSAL_1_1, nullptr, tc.before);
   ASSERT_NE(nullptr, module);
   ir::IRContext context(std::move(module), spvtools::MessageConsumer());
-  context.BuildDefUseManager();
+
+  // Force a re-build of def-use manager the next time it's requested.
+  context.InvalidateAnalyses(ir::IRContext::Analysis::kAnalysisDefUse);
```
Comment (s-perron) [inline test/opt/def_use_test.cpp]: this is not testing that KillInst is updating the def-use manger anymore.  We are suppose to build it and have it valid for the call to KillInst.  Then we check the chains at the end.  Now it is building it at the end.

File: source/opt/ir_context.h [hunk 4]
```diff
@@ -25,10 +25,20 @@ namespace ir {
...
  public:
+  // Available analyses.
+  //
+  // When adding a new analysis:
+  //
+  // 1. Make sure it gets invalidated by IRContext methods that add or remove IR
+  //    elements (e.g., KillDef, KillInst, ReplaceAllUsesWith).
+  //
+  // 2. Add handling code in BuildInvalidAnalyses and
+  //    InvalidateAnalysesExceptFor.
```
[the remark this answers is not in the document]
Comment (dnovillo) [reply source/opt/ir_context.h]: Oh!  Of course they have to be powers of 2... silly me.

File: source/opt/ir_context.h (hunk 1 above)
Comment (dnovillo) [reply source/opt/ir_context.h]: Yup.

File: source/opt/ir_context.h (hunk 2 above)
Comment (dnovillo) [reply source/opt/ir_context.h]: Done.

File: test/opt/def_use_test.cpp (hunk 3 above)
Comment (dnovillo) [reply test/opt/def_use_test.cpp]: Ah, yeah.  If I add a call to get_def_use_mgr(), I would restore the original semantics of the test, right?

**Comment by dnovillo** [comment]
I've made the requested changes.  It's still not ready to submit (waiting for #957 to land).

PTAL.

**Comment by dnovillo** [comment]
I've fixed the failures.  I forgot that some instructions will never belong to any basic block.  The method get_instr_block() was blindly de-referencing the map element without checking that.

All tests should pass now.

@s-perron, PTAL.

File: source/opt/ir_context.h [hunk 5]
```diff
@@ -25,11 +25,24 @@ namespace ir {
...
+  //
+  // When adding a new analysis:
+  //
+  // 1. Enum values should be powers of 2. These are cast into uint32_t
+  //    bitmasks, so we can have at most 31 analyses represented.
+  //
+  // 2. Make sure it gets invalidated by IRContext methods that add or remove IR
+  //    elements (e.g., KillDef, KillInst, ReplaceAllUsesWith).
+  //
+  // 4. Add handling code in BuildInvalidAnalyses and
```
Comment (s-perron) [inline source/opt/ir_context.h]: You missed point 3. :)

File: source/opt/ir_context.h [hunk 6]
```diff
@@ -25,11 +25,24 @@ namespace ir {
...
+  //    InvalidateAnalysesExceptFor.
   enum Analysis {
-    kAnalysisNone = 0x0,
-    kAnalysisBegin = 0x1,
+    kAnalysisNone = 0 << 0,
+    kAnalysisBegin = 1 << 0,
     kAnalysisDefUse = kAnalysisBegin,
-    kAnalysisEnd = 0x2
+    kAnalysisInstrToBlockMapping = 1 << 1,
+    kAnalysisEnd = kAnalysisInstrToBlockMapping
```
Comment (s-perron) [inline source/opt/ir_context.h]: I had used kAnalysisEnd as "one-past-the-end".  See the IRContext test.  They do 

 for (Analysis i = IRContext::kAnalysisBegin; i < IRContext::kAnalysisEnd; i <<= 1)

I don't care which way it is, but it something needs to be changed.

Comment (dnovillo) [reply source/opt/ir_context.h]: Ah, yes.  One past the end seems more intuitive.  Fixed.

Also fixed the test AllPreserveFirstOnlyAfterPassWithChange (thanks for the hints!)

File: source/opt/ir_context.h (hunk 5 above)
Comment (dnovillo) [reply source/opt/ir_context.h]: D'oh! :)

**Comment by dnovillo** [comment]
Both issues fixed.  PTAL.

**Comment by dnovillo** [comment]
Also fixed logic for preserving instr-to-block analysis in KillInst (thanks).  PTAL.

**Comment by dnovillo** [comment]
Thanks.  Re-based, re-tested and pushed.

**Pull request closed without merging**